### PR TITLE
dusk-merkle: Handle inconsistent tree paths

### PR DESCRIPTION
**Pull request by HDauven**
## Summary

- make opening and removal traversal fallible when a recorded position has a missing root or mid-path node
- reject positions at or beyond capacity before target-dependent index conversion
- preserve recorded positions, length, and root when removal is rejected
- add regressions for missing paths, wasm32-like truncation, and no mutation

## Validation

- `make test`
- `make no-std`
- `make clippy`
- `make fmt CHECK=1`
- `make check`
- tree regressions with overflow checks enabled

## Compatibility and residual work

The public API and archive layout are unchanged, and valid tree operations retain their existing behavior. Semantic validation of archived paths, positions, and item-less leaves is intentionally deferred to #119; in particular, lazy cache population is not treated here as a durable archived-leaf invariant.

Refs #110

### CODE REVIEW

File: dusk-merkle/src/node.rs [hunk 1]
```diff
@@ -74,6 +74,18 @@ where
...
+    pub(crate) fn contains_path(&self, height: usize, position: u64) -> bool {
+        if height == H {
+            return self.item.borrow().is_some();
+        }
+
+        let (child_index, child_pos) = Self::child_location(height, position);
+        self.children
+            .get(child_index)
+            .and_then(Option::as_deref)
+            .is_some_and(|child| child.contains_path(height + 1, child_pos))
```
Comment (moCello) [inline dusk-merkle/src/node.rs:86]: **BLOCKER:** Out-of-capacity positions are rejected only as a side effect of `child_location`'s `usize` cast landing outside `[0, A)` — which fails on 32-bit targets, where the cast truncates.

**Context:** `child_location` computes `child_index = (position / child_cap) as usize` under `#[allow(clippy::cast_possible_truncation)]`. `contains_path` rejects an out-of-capacity position solely because that index falls outside `[0, A)` and `get` returns `None`. The two are equivalent only where the cast is lossless. On a 32-bit `usize` the quotient wraps modulo 2^32 whenever `A^(H-1) < 2^32`, and an out-of-capacity position maps back onto a *valid* child index.

**Impact:** With `H = 3, A = 2` on a 32-bit target, `position = 2^34` gives a quotient of `2^32`, truncated to child index `0`. `contains_path` walks leaf 0's path and returns `true`, so `remove(2^34)` takes leaf 0's item, prunes its node, and drops `2^34` from `positions` — destroying a valid leaf, silently moving the root, and leaving position `0` recorded with no path, which is the exact inconsistency this PR exists to prevent. `opening(2^34)` hands back leaf 0's proof. Piecrust's page trees (`H = 8, A = 4`) truncate the same way. No consumer is exploitable today — rusk's note tree is `H = 17, A = 4`, where `4^16 == 2^32` caps the quotient at `u32::MAX` — but the transfer contract compiles this crate to wasm32, so a 32-bit `usize` is a production target and `H`/`A` are free type parameters.

**Severity-rationale:** #110 requires out-of-capacity positions to return `None`. As written that holds only for `H`/`A`/target combinations where the cast is lossless — a property of the caller's configuration, not of this code. `out_of_bounds_tree_position_returns_none` uses `position = capacity()`, which never truncates, so the thumbv6m build does not cover it.

**Fix:** Mirror the bound `Tree::insert` already asserts: add `if position >= self.capacity() { return None; }` to `Tree::remove` and `Tree::opening` ahead of `contains_path`. O(1), independent of the cast, and it leaves `contains_path` answering only the missing-path question.

File: dusk-merkle/src/node.rs [hunk 2]
```diff
@@ -74,6 +74,18 @@ where
         (child_index, child_pos)
     }
 
+    pub(crate) fn contains_path(&self, height: usize, position: u64) -> bool {
+        if height == H {
+            return self.item.borrow().is_some();
+        }
```
Comment (moCello) [inline dusk-merkle/src/node.rs:80]: **WARNING:** The leaf check is order-dependent — `Node::item()` lazily fills a childless node with `EMPTY_SUBTREE`, so an unpopulated leaf reports `true` once anything has computed the root.

**Context:** `Node::item()` recurses into every existing child and, for a childless node, does `self.item.replace(Some(T::EMPTY_SUBTREE))`. A leaf whose path exists but whose `item` is `None` — the state a corrupt archive produces, and the one this arm is meant to reject — is childless, so any prior `Tree::root()`, `smallest_subtree()`, `walk()`, or `opening()` on a *different* position permanently flips its `item` to `Some`.

**Impact:** `opening(p)` returns `None` on a freshly deserialized tree and `Some` once `root()` has been called — same tree, same position, different answer. `remove(p)` likewise flips from `None` to `Some(EMPTY_SUBTREE)` and mutates the tree. `item.is_some()` is a cache-population fact, not an "a value was inserted here" fact.

**Severity-rationale:** Not a blocker — an archive that can supply an item-less leaf can equally supply one holding arbitrary data, so this opens no surface a fully-controlled archive does not already have. It is still a hole in the guard, and a read callers expect to be pure returning different answers depending on call order.

**Fix:** The arm needs a fact the lazy cache cannot forge. Either:

- Reject an item-less leaf during archive validation — `insert`/`remove` cannot produce one (`Node::remove` prunes an emptied leaf's node outright), so `rkyv` is the only source.
- Or make `Node::remove` return `Option<(T, bool)>` and `fill_opening` fallible, so the leaf's real state decides at the point of use.

Returning `true` unconditionally at height `H` is not an option: `Node::remove` does `self.item.take().unwrap()` on the leaf, and this check is what currently keeps that unwrap alive.

File: dusk-merkle/src/tree.rs [hunk 3]
```diff
@@ -225,6 +229,31 @@ mod tests {
...
+        let position = 5;
+        tree.insert(position, 42);
+
+        let (child_index, _) = Node::<u8, H, A>::child_location(0, position);
+        tree.root.children[child_index] = None;
+
+        assert!(tree.opening(position).is_none());
+        assert!(tree.remove(position).is_none());
+        assert!(tree.contains(position));
+    }
```
Comment (moCello) [inline dusk-merkle/src/tree.rs:276]: **WARNING:** Neither regression test asserts the tree was left unmutated after the rejected `remove` — an explicit acceptance criterion of #110 — and only the first level of the recursive descent is exercised.

**Context:**

- `contains(position)` shows the position is still recorded and says nothing about the node tree or the root, so the "no mutation" criterion is untested in both tests.
- Only the root's direct child is nulled, so `contains_path` returns `false` on its first recursion. A node missing *mid-path* (level 1 present, level 2 absent) — the case that depends on the recursion being correct — never runs.

**Fix:** Assert `len()` and `*tree.root()` are unchanged across the failed `remove` in both tests, and null a deeper node in a third case.

File: dusk-merkle/src/tree.rs [hunk 4]
```diff
@@ -63,7 +63,9 @@ where
     /// Remove and return the item at the given `position` in the tree if it
     /// exists.
     pub fn remove(&mut self, position: u64) -> Option<T> {
-        if !self.positions.contains(&position) {
+        if !self.positions.contains(&position)
+            || !self.root.contains_path(0, position)
+        {
             return None;
         }
```
Comment (moCello) [inline dusk-merkle/src/tree.rs]: **NOTE:** A recorded position with no path is now permanently unremovable, and `contains`/`len` keep reporting it while `opening`/`remove` refuse to serve it.

**Context:** `remove` bails out before pruning the bogus entry — correctly, since #110 requires that a failed removal not mutate the tree — and `contains`/`len` still read `positions` alone. A consumer holding a tree recovered from a corrupt archive can then neither detect nor repair the divergence: `len()` over-counts forever, and `contains(p) == true` no longer implies `opening(p).is_some()`.

**Fix:** State the contract, or remove the need for one:

- Document on `contains`, `len`, `opening` and `remove` that a position may be recorded without a populated path, and what each method does in that case.
- Better: validate the invariant once during archive validation. `rkyv` is the only route into this state — `insert`/`remove` keep `positions` and the node path in lockstep, and `Node::children` is `pub(crate)` — so checking it there dissolves the unrepairable entry, both read-path guards, and their per-call cost together.

File: dusk-merkle/src/node.rs [hunk 5]
```diff
@@ -74,6 +74,18 @@ where
         (child_index, child_pos)
     }
 
+    pub(crate) fn contains_path(&self, height: usize, position: u64) -> bool {
```
Comment (moCello) [inline dusk-merkle/src/node.rs:77]: **NOTE:** The panicking internals are fenced off by caller discipline plus a redundant second descent, rather than made unreachable.

**Context:** `Node::remove` (`expect("There should be a child at this position")`, `self.item.take().unwrap()`) and `fill_opening` (`node.children[child_index]` indexing, `expect`) still panic on an inconsistent path. Their safety now rests on every caller having first called `contains_path` along the *identical* index chain; nothing enforces that, so a new call site or a drift between the two descents reintroduces the panic. It also costs an extra O(H) traversal on every valid `opening`/`remove`, paid even in builds without `rkyv-impl`, where the state being defended against cannot occur.

**Fix:** Make `Node::remove` return `Option<(T, bool)>` and `fill_opening` fallible, so one traversal both validates and acts and the panics leave the internals. If the two-pass shape stays, document on `contains_path` that it is the precondition those two rely on, so the coupling is visible at both ends.

**Review by moCello — CHANGES_REQUESTED** [review]
REQUEST CHANGES — the missing-node guard is right, but the other two jobs `contains_path` is doing it does not reliably do. Out-of-capacity rejection leans on `child_location`'s `usize` cast landing out of range, which only holds where that cast is lossless; on a 32-bit target with `A^(H-1) < 2^32`, `remove` of an out-of-capacity position destroys the wrong leaf and moves the root. And the leaf check reads an `item` field that `Node::item()` lazily populates behind its back, so `opening` of an unpopulated leaf starts returning `Some` after any root computation. No shipped configuration hits either today, but the transfer contract compiles this crate to wasm32, so 32-bit is a real target and `H`/`A` are the caller's to choose. An explicit `position >= self.capacity()` check makes the first guarantee unconditional and costs nothing.

Beyond that, one question worth settling before this shape hardens: `rkyv` deserialization is the only way into the state being guarded against, so validating the invariant there — rather than at every read, forever — would remove both guards and the unrepairable `positions` entry along with them.

File: dusk-merkle/src/node.rs (hunk 1 above)
Comment (HDauven) [reply dusk-merkle/src/node.rs:86]: Added explicit capacity checks to `remove` and `opening`, and changed the regression to use position `1 << 34` with a valid leaf at position zero, in 6250ca0.

File: dusk-merkle/src/node.rs (hunk 5 above)
Comment (HDauven) [reply dusk-merkle/src/node.rs:77]: Agreed that the private helper creates a two-pass precondition. I am keeping that shape in this targeted fix: making `remove` and `fill_opening` fallible is a broader refactor, and it does not fully solve the lazy-leaf/archive issue by itself. That should be evaluated together with archive validation.

File: dusk-merkle/src/node.rs (hunk 2 above)
Comment (moCello) [inline dusk-merkle/src/node.rs:80]: **WARNING:** The leaf arm tests a lazy cache, not a fact — `Node::item()` fills any childless node with `EMPTY_SUBTREE`, so an item-less leaf reports `true` or `false` depending on the tree's cache state.

**Context:** `Node::item()` recurses into its children only when its own `item` is `None`, and for a childless node it does `self.item.replace(Some(T::EMPTY_SUBTREE))`. So an item-less leaf — the state this arm exists to reject — is flipped to `Some` by any `root()`/`walk()`/`smallest_subtree()`/`opening()` that finds every ancestor on its path uncached. An archive storing those ancestors with `item: None` produces that; so does an ordinary `insert` into a sibling leaf, which clears the shared ancestors and re-aggregates through the bogus leaf.

**Impact:** With a cold cache along the path, `contains_path` returns `true`, and `remove(p)` returns `Some(EMPTY_SUBTREE)`, prunes the path, drops `p` from `positions` and moves the root — a mutating "successful" removal of a leaf that was never populated, which is what #110's "failed removal does not mutate the tree" criterion excludes. The same tree answers `None` if the cache happens to be warm. The actor who can produce the inconsistency also controls the cached items, so the check can be defeated deliberately.

**Severity-rationale:** Not a blocker — the #110 panic (`item.take().unwrap()`) is genuinely gone, and reaching this state needs a corrupt archive, which could equally supply a leaf holding arbitrary data. What survives is a guard that does not reliably guard, plus a silent tree mutation. Whether that ships under a PR saying "Closes #110" is a maintainer judgment call.

**Fix:** The arm needs a fact the lazy cache cannot forge. Either reject an item-less leaf during archive validation (`insert`/`remove` cannot produce one — `Node::remove` prunes an emptied leaf's node outright — so `rkyv` is the only source), or make `Node::remove` return `Option<(T, bool)>` and `fill_opening` fallible so the leaf's real state decides at the point of use. Returning `true` unconditionally at height `H` is not an option: `Node::remove` does `self.item.take().unwrap()` on the leaf, and this check is what keeps that unwrap alive. Deferring to the archive-validation change you described is reasonable, provided it is tracked in an issue and #110 is not closed as fully handled here.

This arm is also the one new branch with no test — both path tests null a child pointer and so exercise the recursion arm, and the out-of-bounds test never reaches `contains_path`. For the regression test, graft `Node::new()` down the full path to the leaf (`pub(crate)`, so reachable from the `tree.rs` test module), record the position, and assert `opening`/`remove` return `None` both before and after a `tree.root()` call. The second assertion is the one that fails today.

File: dusk-merkle/src/tree.rs [hunk 6]
```diff
@@ -63,7 +63,10 @@ where
     /// Remove and return the item at the given `position` in the tree if it
     /// exists.
     pub fn remove(&mut self, position: u64) -> Option<T> {
-        if !self.positions.contains(&position) {
+        if position >= self.capacity()
```
Comment (moCello) [inline dusk-merkle/src/tree.rs:66]: **NOTE:** The PR introduces a "recorded but unservable" state that no doc-comment mentions — `contains(p) == true` no longer implies `opening(p).is_some()`.

**Context:** `remove(p)` cannot clear the bogus entry — it bails out before pruning, correctly, since a failed removal must not mutate the tree — while `contains` and `len` still read `positions` alone, so `len()` keeps over-counting. The only route back to a consistent tree is `insert(p, x)` followed by `remove(p)`, and nothing in the API says so, or that the divergence exists at all: `None` from `remove`/`opening` is indistinguishable from "position absent". `remove` and `opening` still say only "if it exists"; `contains`/`len` still describe pure `positions` semantics.

**Fix:** Either state the contract on `contains`, `len`, `opening` and `remove` — that a position may be recorded without a populated path, and what each method does in that case — or dissolve the need for one by validating the invariant during archive validation. `rkyv` is the only route into this state, since `insert`/`remove` keep `positions` and the node path in lockstep and `Node::children` is `pub(crate)`.

File: dusk-merkle/src/tree.rs [hunk 7]
```diff
@@ -78,7 +81,10 @@ where
     where
         T: Clone,
     {
-        if !self.positions.contains(&position) {
+        if position >= self.capacity()
+            || !self.positions.contains(&position)
+            || !self.root.contains_path(0, position)
+        {
             return None;
         }
```
Comment (moCello) [inline dusk-merkle/src/tree.rs:89]: **NOTE:** The three-condition guard is duplicated verbatim in `remove` and `opening`, and nothing at either site records that it is now the precondition the panicking internals rely on.

**Context:** `Node::remove` (`expect("There should be a child at this position")`, `self.item.take().unwrap()`) and `fill_opening` (`node.children[child_index]` indexing, `expect`) stay panicking; their safety rests on every caller having first run `contains_path` along the identical index chain. The two copies can drift independently.

**Fix:** Fold the guard into one private `Tree` helper (e.g. `fn has_populated_leaf(&self, position: u64) -> bool`) called from both sites, and document on it that `Node::remove` and `fill_opening` depend on it holding. That gives the coupling a single visible home until the fallible-internals refactor lands.

**Review by moCello — COMMENTED** [review]
COMMENT — the previous blocker is properly fixed: the explicit `position >= self.capacity()` check makes out-of-capacity rejection unconditional rather than a by-product of the `usize` cast, and the regression tests now pin both the no-mutation criterion and a mid-path missing node. The panic #110 is about is gone, so this does not block.

One thing to settle before merge: the guard covers two of the three inconsistency shapes. An item-less leaf whose path exists slips through whenever the lazy item cache is cold along that path, and `remove` then mutates the tree and returns a phantom item. Deferring the fallible-internals refactor to the archive-validation change is a sound call, but please open an issue for it and drop "Closes #110" from this PR, or take the narrow fix now — #110's acceptance criteria are not all met by the diff as it stands.

File: dusk-merkle/src/node.rs (hunk 2 above)
Comment (HDauven) [reply dusk-merkle/src/node.rs]: Deferred to https://github.com/dusk-network/merkle/issues/119; this PR now references rather than closes #110.

**Review by moCello — APPROVED** [review]
APPROVE — the fallible-internals refactor lands what earlier rounds asked for: `Node::remove` and `fill_opening` return `Option` so one traversal both validates and acts, the panic sites are gone from the removal and opening paths, out-of-capacity rejection is a capacity check in a single shared helper that makes the index cast in `child_location` safe by construction, and every failure path leaves the tree unmutated — lazy hash caches included — with regression tests pinning the root-level, mid-path, and out-of-capacity cases. The remaining gap — an item-less archived leaf answered from the lazily-warmable item cache, and the recorded-but-unservable positions that come with it — is tracked in #119, and the PR now references rather than closes #110. Nothing here asks for a change on this branch.
